src/main/java/apoc/index/IndexUpdateTransactionEventHandler.java

```diff
@@ -0,0 +1,197 @@
+            index.remove(node, FreeTextSearch.KEY);
+        }));
+
+        iterateLabelChanges(data.assignedLabels(), data.createdNodes(), (index, node, key, value, ignore) -> indexUpdate(state, aVoid -> {
+            index.add(node, key, value);
+            index.add(node, FreeTextSearch.KEY, value);
+        }));
+
+        iterateLabelChanges(data.removedLabels(), data.deletedNodes(), (index, node, key, value, ignore) -> indexUpdate(state, aVoid -> {
+            index.remove(node, key);
+            index.remove(node, FreeTextSearch.KEY);
+        }));
+
+        return state;
+    }
+
+    @Override
+    public void afterCommit(TransactionData data, Collection<Consumer<Void>> state) {
+        if (async) {
+            for (Consumer<Void> consumer: state) {
+                try {
+                    indexCommandQueue.put(consumer);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+    }
+
+    private void iterateNodePropertyChange(Iterable<PropertyEntry<Node>> propertyEntries, IndexFunction<Index<Node>, Node, String, Object, Object> function) {
+        propertyEntries.forEach(nodePropertyEntry -> {
+            final Node entity = nodePropertyEntry.entity();
+            final String key = nodePropertyEntry.key();
+            final Object value = nodePropertyEntry.value();
+            Collection<Index<Node>> indices = findIndicesAffectedBy(entity.getLabels(), key);
+            for (Index<Node> index : indices) {
+                function.apply(index, entity, key, value, nodePropertyEntry.previouslyCommitedValue());
+            }
+        });
+    }
+
+    private void iterateLabelChanges(Iterable<LabelEntry> labelEntries, Iterable<Node> blacklist, IndexFunction<Index<Node>, Node, String, Object, Void> function) {
+        StreamSupport.stream(labelEntries.spliterator(), false)
+                .filter(labelEntry -> iterableContains(blacklist, labelEntry.node()))
+                .forEach(labelEntry -> {
+                    final Map<String, Collection<Index<Node>>> propertyIndicesMap = indexesByLabelAndProperty.get(labelEntry.label().name());
+                    if (propertyIndicesMap!=null) {
+                        final Node entity = labelEntry.node();
+                        for (String key : entity.getPropertyKeys()) {
+                            Collection<Index<Node>> indices = propertyIndicesMap.get(key);
+                            if (indices != null) {
+                                for (Index<Node> index : indices) {
+                                    Object value = entity.getProperty(key);
+                                    function.apply(index, entity, key, value, null);
+                                }
+                            }
+                        }
+                    }
+                });
+    }
+
+    private boolean iterableContains(Iterable<Node> nodes, Node node) {
```

Comment:
I'd probably convert the iterable into a set once before this call and do a check against the set.